File: src/python3/development/hardware_classes/drivetrain.py
from imu import IMU
import RPi.GPIO as GPIO
from time import sleep


# Drivetrain constants
LEFT_MOTOR_DIR_PIN = 16
LEFT_MOTOR_STEP_PIN = 20
RIGHT_MOTOR_DIR_PIN = 19
RIGHT_MOTOR_STEP_PIN = 26
SLEEP_PIN = 21

SPR = 800       		# Signal pulses Per Revolution
DRIVING_DELAY = 0.0001	# Time between signal pulses when driving
TURNING_DELAY = 0.0005	# Time between signal pulses when turning


class Drivetrain:
	def __init__(self, dir_pin_1 = LEFT_MOTOR_DIR_PIN, step_pin_1 = LEFT_MOTOR_STEP_PIN, \
					dir_pin_2 = RIGHT_MOTOR_DIR_PIN, step_pin_2 = RIGHT_MOTOR_STEP_PIN, sleep_pin = SLEEP_PIN, imu_auto_calibrate = False, debug = False):
		self.__debug = debug

		# Left motor setup
		self.__dir_pin_1 = dir_pin_1
		self.__step_pin_1 = step_pin_1

		# Right motor setup
		self.__dir_pin_2 = dir_pin_2
		self.__step_pin_2 = step_pin_2

		# Sleep pin
		self.__sleep_pin = sleep_pin

		# GPIO setup
		GPIO.setwarnings(False)
		GPIO.setmode(GPIO.BCM)

		GPIO.setup(self.__dir_pin_1, GPIO.OUT)
		GPIO.setup(self.__step_pin_1, GPIO.OUT)
		GPIO.setup(self.__dir_pin_2, GPIO.OUT)
		GPIO.setup(self.__step_pin_2, GPIO.OUT)
		GPIO.setup(self.__sleep_pin, GPIO.OUT)

		self.__imu = IMU(auto_calibrate = imu_auto_calibrate, debug = debug)


	def rotate(self, direction):
		if direction == 'f':
			GPIO.output(self.__dir_pin_1, GPIO.HIGH)
			GPIO.output(self.__dir_pin_2, GPIO.HIGH)
		elif direction == 'b':
			GPIO.output(self.__dir_pin_1, GPIO.LOW)
			GPIO.output(self.__dir_pin_2, GPIO.LOW)

		GPIO.output(self.__step_pin_1, GPIO.HIGH)
		GPIO.output(self.__step_pin_2, GPIO.HIGH)
		sleep(DRIVING_DELAY)

		GPIO.output(self.__step_pin_1, GPIO.LOW)
		GPIO.output(self.__step_pin_2, GPIO.LOW)
		sleep(DRIVING_DELAY)


	def drive(self, direction, rotations):
		if direction == 'f':
			GPIO.output(self.__dir_pin_1, GPIO.HIGH)
			GPIO.output(self.__dir_pin_2, GPIO.HIGH)
		elif direction == 'b':
			GPIO.output(self.__dir_pin_1, GPIO.LOW)
			GPIO.output(self.__dir_pin_2, GPIO.LOW)
	
		for i in range(rotations * SPR):
			GPIO.output(self.__step_pin_1, GPIO.HIGH)
			GPIO.output(self.__step_pin_2, GPIO.HIGH)

			sleep(DRIVING_DELAY)

			GPIO.output(self.__step_pin_1, GPIO.LOW)
			GPIO.output(self.__step_pin_2, GPIO.LOW)
			
			sleep(DRIVING_DELAY)


	# Works for turns up to 360 degrees
	def turn(self, direction, degrees):
		turning_offset = (degrees % 360.0) / 30

		if direction == 'l':
			GPIO.output(self.__dir_pin_1, GPIO.LOW)
			GPIO.output(self.__dir_pin_2, GPIO.HIGH)

			final_orientation = self.__imu.get_yaw_value() - degrees % 360.0
			turning_offset *= -1

			if final_orientation < 0:
				final_orientation = 360.0 + final_orientation

		elif direction == 'r':
			GPIO.output(self.__dir_pin_1, GPIO.HIGH)
			GPIO.output(self.__dir_pin_2, GPIO.LOW)

			final_orientation = (self.__imu.get_yaw_value() + degrees % 360.0) % 360.0

		while round(self.__imu.get_yaw_value(), 0) != round(final_orientation - turning_offset, 0):
			GPIO.output(self.__step_pin_1, GPIO.HIGH)
			GPIO.output(self.__step_pin_2, GPIO.HIGH)

			sleep(TURNING_DELAY)
			
			GPIO.output(self.__step_pin_1, GPIO.LOW)
			GPIO.output(self.__step_pin_2, GPIO.LOW)
			
			sleep(TURNING_DELAY)


	def toggle_power(self, is_on):
		GPIO.output(self.__sleep_pin, is_on)

File: src/development/cpp/drivetrain.cpp
#include "drivetrain.h"

unsigned short SPR = 800; // Signal pulses Per Revolution.
unsigned int DRIVING_DELAY = 500; // Microseconds (0.0005 seconds). Time between signal pulses when driving.
unsigned int TURNING_DELAY = 500; // Microseconds (0.0005 seconds). Time between signal pulses when turning.

Drivetrain::Drivetrain(unsigned short left_motor_dir_pin, unsigned short left_motor_step_pin,
                        unsigned short right_motor_dir_pin, unsigned short right_motor_step_pin) 
{
    this->left_motor_dir_pin = left_motor_dir_pin;
    this->left_motor_step_pin = left_motor_step_pin;
    this->right_motor_dir_pin = right_motor_dir_pin;
    this->right_motor_step_pin = right_motor_step_pin;

    wiringPiSetupGpio();

    pinMode(left_motor_dir_pin, OUTPUT);
    pinMode(left_motor_step_pin, OUTPUT);
    pinMode(right_motor_dir_pin, OUTPUT);
    pinMode(right_motor_step_pin, OUTPUT);
}

// Direction 'f' = forwards, 'b' = backwards.
// Should be used in a while loop.
void Drivetrain::drive(char direction)
{   
    if(direction == 'f')
    {
        digitalWrite(left_motor_dir_pin, HIGH);
        digitalWrite(right_motor_dir_pin, HIGH);
    }
    else if(direction == 'b')
    {
        digitalWrite(left_motor_dir_pin, LOW);
        digitalWrite(right_motor_dir_pin, LOW);
    }

    digitalWrite(left_motor_step_pin, HIGH);
    digitalWrite(right_motor_step_pin, HIGH);

    usleep(DRIVING_DELAY);

    digitalWrite(left_motor_step_pin, LOW);
    digitalWrite(right_motor_step_pin, LOW);

    usleep(DRIVING_DELAY);
}

// Direction 'l' = left, 'r' = right. 
void Drivetrain::turn(float rotations, char direction)
{
    if(direction == 'l')
    {
        digitalWrite(left_motor_dir_pin, LOW);
        digitalWrite(right_motor_dir_pin, HIGH);
    }
    else if(direction == 'r')
    {
        digitalWrite(left_motor_dir_pin, HIGH);
        digitalWrite(right_motor_dir_pin, LOW);
    }

    for(int pulse = 0; pulse < (int) SPR * rotations; pulse++)
    {
        digitalWrite(left_motor_step_pin, HIGH);
        digitalWrite(right_motor_step_pin, HIGH);

        usleep(TURNING_DELAY);

        digitalWrite(left_motor_step_pin, LOW);
        digitalWrite(right_motor_step_pin, LOW);

        usleep(TURNING_DELAY);
    }
}

// Direction 'f' = forwards, 'b' = backwards.
void Drivetrain::rotate(float rotations, char direction)
{
    if(direction == 'f')
    {
        digitalWrite(left_motor_dir_pin, HIGH);
        digitalWrite(right_motor_dir_pin, HIGH);
    }
    else if(direction == 'b')
    {
        digitalWrite(left_motor_dir_pin, LOW);
        digitalWrite(right_motor_dir_pin, LOW);
    }

    for(int pulse = 0; pulse < (int) SPR * rotations; pulse++)
    {
        digitalWrite(left_motor_step_pin, HIGH);
        digitalWrite(right_motor_step_pin, HIGH);

        usleep(DRIVING_DELAY);

        digitalWrite(left_motor_step_pin, LOW);
        digitalWrite(right_motor_step_pin, LOW);

        usleep(DRIVING_DELAY);
    }
}
